File: attractor/agent/loop_detector.py
"""Loop detection for the agent loop."""

from __future__ import annotations

import hashlib
import json
from collections import deque

from pydantic import BaseModel


class LoopDetection(BaseModel):
    is_looping: bool = False
    description: str = ""
    pattern_length: int = 0


class LoopDetector:
    """Detects when the agent is stuck in a loop calling the same tools."""

    def __init__(self, window_size: int = 20, threshold: int = 3):
        self._window: deque[str] = deque(maxlen=window_size)
        self._threshold = threshold

    def record(self, tool_name: str, args: dict) -> LoopDetection:
        """Record a tool call and check for loops."""
        # Create a hash of tool name + args
        sig = self._signature(tool_name, args)
        self._window.append(sig)

        # Check for repeating patterns
        window = list(self._window)
        n = len(window)

        # Check for same call repeated threshold times
        if n >= self._threshold:
            last = window[-1]
            consecutive = sum(
                1 for i in range(n - 1, -1, -1) if window[i] == last
            )
            if consecutive >= self._threshold:
                return LoopDetection(
                    is_looping=True,
                    description=f"Same tool call repeated {consecutive} times: {tool_name}",
                    pattern_length=1,
                )

        # Check for repeating sequences (length 2 and 3)
        for pattern_len in (2, 3):
            if n >= pattern_len * self._threshold:
                pattern = window[-pattern_len:]
                repeats = 0
                for i in range(self._threshold):
                    start = n - pattern_len * (i + 1)
                    if start < 0:
                        break
                    segment = window[start:start + pattern_len]
                    if segment == pattern:
                        repeats += 1
                if repeats >= self._threshold:
                    return LoopDetection(
                        is_looping=True,
                        description=f"Repeating pattern of {pattern_len} calls detected",
                        pattern_length=pattern_len,
                    )

        return LoopDetection()

    def reset(self) -> None:
        self._window.clear()

    @staticmethod
    def _signature(tool_name: str, args: dict) -> str:
        args_str = json.dumps(args, sort_keys=True, default=str)
        h = hashlib.md5(args_str.encode()).hexdigest()[:8]
        return f"{tool_name}:{h}"

File: internal/agent/loop_detector.go
package agent

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
)

// LoopDetection reports whether the agent appears stuck.
type LoopDetection struct {
	IsLooping     bool
	Description   string
	PatternLength int
}

// LoopDetector spots repeated tool-call patterns within a sliding window.
type LoopDetector struct {
	window     []string
	windowSize int
	threshold  int
}

// NewLoopDetector returns a detector with the given window size and threshold.
func NewLoopDetector(windowSize, threshold int) *LoopDetector {
	if windowSize <= 0 {
		windowSize = 20
	}
	if threshold <= 0 {
		threshold = 3
	}
	return &LoopDetector{windowSize: windowSize, threshold: threshold}
}

// Record logs a tool call and checks for loops.
func (d *LoopDetector) Record(toolName string, args map[string]any) LoopDetection {
	sig := signature(toolName, args)
	d.window = append(d.window, sig)
	if len(d.window) > d.windowSize {
		d.window = d.window[len(d.window)-d.windowSize:]
	}

	w := d.window
	n := len(w)

	// Same call repeated threshold times consecutively.
	if n >= d.threshold {
		last := w[n-1]
		consecutive := 0
		for i := n - 1; i >= 0; i-- {
			if w[i] == last {
				consecutive++
			} else {
				break
			}
		}
		if consecutive >= d.threshold {
			return LoopDetection{IsLooping: true, PatternLength: 1,
				Description: fmt.Sprintf("Same tool call repeated %d times: %s", consecutive, toolName)}
		}
	}

	// Repeating sequences of length 2 and 3.
	for _, patternLen := range []int{2, 3} {
		if n >= patternLen*d.threshold {
			pattern := w[n-patternLen:]
			repeats := 0
			for i := 0; i < d.threshold; i++ {
				start := n - patternLen*(i+1)
				if start < 0 {
					break
				}
				if equalSlice(w[start:start+patternLen], pattern) {
					repeats++
				}
			}
			if repeats >= d.threshold {
				return LoopDetection{IsLooping: true, PatternLength: patternLen,
					Description: fmt.Sprintf("Repeating pattern of %d calls detected", patternLen)}
			}
		}
	}
	return LoopDetection{}
}

// Reset clears the detection window.
func (d *LoopDetector) Reset() { d.window = nil }

func signature(toolName string, args map[string]any) string {
	b, _ := json.Marshal(args)
	sum := md5.Sum(b)
	return toolName + ":" + hex.EncodeToString(sum[:])[:8]
}

func equalSlice(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
